cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/distributed/runtimejob/DefaultRuntimeJob.java

```diff
@@ -657,15 +549,155 @@ private void stopCoreServices(Deque<Service> coreServices,
-            e);
+        LOG.warn(
+            "Exception raised when stopping service {} during program termination.",
+            service, e);
+      }
+    }
+  }
+
+  /**
+   * Regenerates the plugin artifacts and returns the new
+   * {@link ApplicationSpecification} and {@link ProgramOptions}.
+   */
+  private ImmutablePair<ApplicationSpecification, ProgramOptions> regeneratePluginArtifacts(
+      final ProgramOptions originalProgramOpts,
+      final ProgramDescriptor programDescriptor, final Injector injector)
+      throws Exception {
+    ConfiguratorFactory configuratorFactory = injector.getInstance(
+        ConfiguratorFactory.class);
+    CConfiguration cConf = injector.getInstance(CConfiguration.class);
+
+    ApplicationSpecification appSpec = programDescriptor.getApplicationSpecification();
+    ProgramId programId = programDescriptor.getProgramId();
+    Map<String, String> systemArguments = new HashMap<>(
+        originalProgramOpts.getArguments().asMap());
+    File pluginDir = new File(originalProgramOpts.getArguments()
+        .getOption(ProgramOptionConstants.PLUGIN_DIR,
+            ProgramRunners.PLUGIN_DIR));
+    // create a directory to store plugin artifacts for the regeneration of app spec to fetch plugin artifacts
+    DirUtils.mkdirs(pluginDir);
+
+    if (!originalProgramOpts.getArguments()
+        .hasOption(ProgramOptionConstants.PLUGIN_DIR)) {
+      systemArguments.put(ProgramOptionConstants.PLUGIN_DIR,
+          ProgramRunners.PLUGIN_DIR);
+    }
+
+    // remember the file names in the artifact folder before app regeneration
+    final List<String> pluginFiles = DirUtils.listFiles(pluginDir, File::isFile)
+        .stream().map(File::getName).collect(Collectors.toList());
+
+    ApplicationSpecification newAppSpec = regenerateAppSpec(systemArguments,
+        originalProgramOpts.getUserArguments().asMap(), programId, appSpec,
+        programDescriptor, configuratorFactory);
+    appSpec = newAppSpec != null ? newAppSpec : appSpec;
+
+    List<String> pluginFilesAfter = DirUtils.listFiles(pluginDir, File::isFile)
+        .stream().map(File::getName).collect(Collectors.toList());
+
+    if (pluginFilesAfter.isEmpty()) {
+      systemArguments.remove(ProgramOptionConstants.PLUGIN_DIR);
+    }
+
+    // if different, create an updated plugin archive
+    if (!pluginFiles.equals(pluginFilesAfter)) {
+      File tempDir = DirUtils.createTempDir(
+          new File(cConf.get(Constants.CFG_LOCAL_DATA_DIR),
+              cConf.get(Constants.AppFabric.TEMP_DIR)).getAbsoluteFile());
+      File tempArchiveDir = DirUtils.createTempDir(tempDir);
+      File rebuiltPluginArchive = ProgramRunners.createPluginArchive(
+          originalProgramOpts, tempArchiveDir);
+      systemArguments.put(ProgramOptionConstants.PLUGIN_ARCHIVE,
+          rebuiltPluginArchive.getAbsolutePath());
+    }
+
+    // update program options
```

Comment:
nit: from this method perspective its "generate" not "update"

Reply:
Reworded.